### ISSUE TITLE
There might be a bug when generating the fused features.

### ISSUE THREAD

**Issue by dongyibo** [NickL77/BaldEagle#18]
`
with torch.no_grad():
        outputs = model(
            row["input_ids"].unsqueeze(0)[:, :max_token_length].cuda(),
            output_hidden_states=True,
        )
        # hidden_states = outputs.hidden_states[-1].cpu()
        low_layer = outputs.hidden_states[low_layer_idx].cpu()
        mid_layer = outputs.hidden_states[mid_layer_idx].cpu()
        high_layer = outputs.hidden_states[high_layer_idx].cpu()
        hidden_states = torch.concat([low_layer, mid_layer, high_layer], dim=2)
        target_hidden_states = outputs.hidden_states[-1].cpu()
    data_point = {
        "input_ids": row["input_ids"],
        "loss_mask": row["loss_mask"],
        "hidden_state": hidden_states,
        "target_hidden_states": target_hidden_states,
    }
`

I notice that: length of outputs.hidden_states is {num_hidden_layers+1}, The extra layer should be the word embedding layer，so the index of low/mid/high show + 1 ？

**Comment by NickL77** [comment]
I took a look at this previously. If you look at sglang's implementation, I believe their 0-index hidden_state is also the token_embedding as when they check to save, it's before they call the first decoder layer: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/llama.py#L329-L339

So in both HuggingFace and Sglang, 0 layer is embedding, 1 layer is first decoder output, etc. Does this seem right to you?

**Comment by dongyibo** [comment]
> I took a look at this previously. If you look at sglang's implementation, I believe their 0-index hidden_state is also the token_embedding as when they check to save, it's before they call the first decoder layer: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/llama.py#L329-L339
> 
> So in both HuggingFace and Sglang, 0 layer is embedding, 1 layer is first decoder output, etc. Does this seem right to you?

-------
I saw eagle3 train code from Eagle Team:
https://github.com/SafeAILab/EAGLE/blame/c9d050a641f29b9b882cf9e1c1eedafd5e6e51df/eagle/traineagle3/modeling_llama_kv.py#L1137
and:
---
self.layers = nn.ModuleList(
            [LlamaDecoderLayer(config) for _ in range(config.num_hidden_layers)]
        )
---
where config.num_hidden_layers not includes embedding layer.
so I think 0 layer is  first decoder output from Eagle Team‘s perspective

**Comment by dongyibo** [comment]
> https://github.com/SafeAILab/EAGLE/blame/c9d050a641f29b9b882cf9e1c1eedafd5e6e51df/eagle/traineagle3/modeling_llama_kv.py#L1123-L1140C17
> 
> Line 1123 assigns hidden_states to inputs_embeds. Between Lines 1123 and 1140, we don't run any decoder layer. And on line 1140, we save hidden_states, so hidden_states[0] is just the token embeddings. The decoder in only run on line 1155 or 1163. So the output of the first decoder layer is stored only after the for loop iterates through once and is set as hidden_states[1].

You inspired me, thank you.
